lib/src/SlashRemover.cc

```diff
@@ -73,31 +75,53 @@ void SlashRemover::initAndStart(const Json::Value& config)
         (trailingSlashes_ * trailing) | (duplicateSlashes_ * duplicate);
     if (!removeMode)
         return;
-    app().registerHandlerViaRegex(
-        regexes[removeMode - 1],
-        [removeMode,
-         this](const HttpRequestPtr& req,
-               std::function<void(const HttpResponsePtr&)>&& callback) {
-            string newPath = req->path();
-            switch (removeMode)
+    auto redirector = app().getPlugin<Redirector>();
+    redirector->registerHandler(
+        [removeMode, redirect = redirect_](const HttpRequestPtr& req,
+                                           std::string& location) {
+            static const std::regex regex(regexes[removeMode - 1]);
+            if (std::regex_match(req->path(), regex))
             {
-                case trailing:
-                    removeTrailingSlashes(newPath);
-                    break;
-                case duplicate:
-                    removeDuplicateSlashes(newPath);
-                    break;
-                case both:
-                default:
-                    removeExcessiveSlashes(newPath);
-                    break;
-            }
-            if (redirect_)
-                callback(HttpResponse::newRedirectionResponse(newPath));
-            else
-            {
-                req->setPath(newPath);
-                app().forward(req, std::move(callback));
+                string newPath = req->path();
+                switch (removeMode)
+                {
+                    case trailing:
+                        removeTrailingSlashes(newPath);
+                        break;
+                    case duplicate:
+                        removeDuplicateSlashes(newPath);
+                        break;
+                    case both:
+                    default:
+                        removeExcessiveSlashes(newPath);
+                        break;
+                }
+                if (redirect)
+                {
+                    req->setPath(newPath);
+                    if (location.empty())
+                    {
+                        location = std::move(newPath);
+                    }
+                    else
+                    {
+                        if (location.find("http") == 0)
+                        {
+                            auto pos = location.find_first_of('/', 8);
+                            if (pos != string::npos)
+                                location.resize(pos);
+                        }
+                        location.append(newPath);
+                        if (!req->query().empty())
+                        {
+                            location.append("?").append(req->query());
+                        }
+                    }
+                }
+                else
+                {
+                    req->setPath(newPath);
```

Comment:
The forwarder is removed, has it been tested whether the request gets forwarded?
```c++
app().forward(req, std::move(callback));
```

Reply:
For this to work properly, we must check if the other redirectors did anything to the location string, then we won't forward, on next request after the forward, the slash remover removes the slashes and sees it is the only redirector that made changes, then it forwards the request

Reply:
The sync join point is at the beginning of the pipeline for all handlers. I think it's ok but I haven't tested it yet.

Reply:
This is the expected behavior with Slash Remover's redirect config set to true

example.com//home///   REDIRECT   https://example.com/home
https://example.com/home   SERVED

Now with redirect set to false

example.com//home///   REDIRECT   https://example.com//home///
https://example.com//home///   FORWARD   https://example.com/home   SERVED